src/coreclr/System.Private.CoreLib/src/System/Runtime/DependentHandle.cs

```diff
@@ -0,0 +1,224 @@
+            _handle = nInitialize(target, dependent);
+        }
+
+        /// <summary>
+        /// Gets a value indicating whether this handle has been allocated or not.
+        /// </summary>
+        public bool IsAllocated => _handle != IntPtr.Zero;
+
+        /// <summary>
+        /// Gets or sets the target object instance for the current handle.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">Thrown if <see cref="IsAllocated"/> is <see langword="false"/>.</exception>
+        public object? Target
+        {
+            get
+            {
+                IntPtr handle = _handle;
+
+                if (handle == IntPtr.Zero)
+                {
+                    ThrowHelper.ThrowInvalidOperationException();
+                }
+
+                return nGetPrimary(handle);
+            }
+            set
+            {
+                IntPtr handle = _handle;
+
+                if (handle == IntPtr.Zero)
+                {
+                    ThrowHelper.ThrowInvalidOperationException();
+                }
+
+                nSetPrimary(handle, value);
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets the dependent object instance for the current handle.
+        /// </summary>
+        /// <remarks>
+        /// If it is necessary to retrieve both <see cref="Target"/> and <see cref="Dependent"/>, it is
+        /// recommended to use <see cref="GetTargetAndDependent"/> instead. This will result in better
+        /// performance and it will reduce the chance of unexpected behavior in some cases.
+        /// </remarks>
+        /// <exception cref="InvalidOperationException">Thrown if <see cref="IsAllocated"/> is <see langword="false"/>.</exception>
+        public object? Dependent
+        {
+            get
+            {
+                IntPtr handle = _handle;
+
+                if (handle == IntPtr.Zero)
+                {
+                    ThrowHelper.ThrowInvalidOperationException();
+                }
+
+                _ = nGetPrimaryAndSecondary(handle, out object? dependent);
+
+                return dependent;
+            }
+            set
+            {
+                IntPtr handle = _handle;
+
+                if (handle == IntPtr.Zero)
+                {
+                    ThrowHelper.ThrowInvalidOperationException();
+                }
+
+                nSetSecondary(handle, value);
+            }
+        }
+
+        /// <summary>
+        /// Retrieves the values of both <see cref="Target"/> and <see cref="Dependent"/>, if available.
+        /// </summary>
+        /// <returns>The values of <see cref="Target"/> and <see cref="Dependent"/>.</returns>
+        /// <exception cref="InvalidOperationException">Thrown if <see cref="IsAllocated"/> is <see langword="false"/>.</exception>
+        public (object? Target, object? Dependent) GetTargetAndDependent()
```

Comment:
I couldn't attend the API review for this... what was the reason this design was decided on rather than:
```C#
public (object? Target, object? Dependent) TargetAndDependent { get; }
```
?

Reply:
Jeremy mentioned that given that the method returned a tuple, it felt much more appropriate to him for it to be a method instead. I agree with that as well, and I'll also add that a tuple-returning property has never been used before in the BCL, whereas a tuple-returning method is more common today (eg. all the new `Math` APIs recently added). Jan said either was fine for him, so I just went with the method in this case 😄

Reply:
> given that the method returned a tuple, it felt much more appropriate to him for it to be a method instead

Why?  We have many properties that return structs with multiple properties.  What makes a ValueTuple special?

Reply:
Can't talk for Jeremy specifically, what he said on stream was:

> "Having a property that returns a tuple feels weird, it feels more method-y"

> "[...] for reasons that I can't articulate, maybe artistic style, this feels more like a method than a property"

> "[...] `TargetAndDependent` should be a method, not a property. It feels really weird as a property".

For more on his part, you'd have to ask him 🙂

On my end, I can say that one of the reasons why I really feel like it should be a method is that in this case we're not conceptually returning an entity (such as a color, which has multiple channels), but two separate entities. The C# guideline often refer to properties as each representing "one entity", which is the case here as well, with `Target` and `Dependent`. As in, we have a property for each entity, and if you want to get both of them together, we then have a method that retrieves both entities and returns them. Having both a property for each entity and also a property for both looks really weird to me. And also the fact that we've just never used a value-tuple in a property before, while we have in methods. I just really agree with Jeremy that a method just feels right in this case, is all 😄

Reply:
Maybe we should just drop this method/property. The motivation for this method was that it helps to avoid certain types of race conditions: https://github.com/dotnet/runtime/issues/19459#issuecomment-854931622 . These race conditions can be trivially avoided by the caller using `GC.KeepAlive` too, so it is there just for convenience.

This type of race conditions should be mentioned in the documentation in any case. Given that this is advanced type, I do not think that there is a huge difference between recomending to use `GC.KeepAlive` vs. recomending to use the tuple returning method.

Reply:
I don't view `Target` and `Dependent` as any different from say `KeyValuePair` which would also be "one entity".

Logically, a value tuple is just a struct, that you can happen give names to inline, and so it shouldn't have "different" rules for when to use a property vs a method.

Reply:
`Enumerator.Current` on most dictionary types returns a `KeyValuePair<TKey, TValue>`; that key and value are equally related as are target/dependent.

`GC.GetMemoryInfo` returns a struct with a bunch of independent data elements.

I guess I'm not seeing the argument (though I appreciate the explanation).  And it's strange to me that Target and Dependent are properties but TargetAndDependent is not.  I'm not going to die on that hill, but, yeah.

Reply:
To be clear, as I mentioned to Tanner already, that for me was just a minor style preference, but I'd be absolutely fine with a property as well. For me really the only important bit is to have a single API to get both and avoid doing two internal calls there (as in, the smaller overhead here would be another advantage of having this API other than it being less error prone than accessing `Target` and `Dependent` separately, if you eg. forget to use `GC.KeepAlive`). The fact that it's a property or a method is mostly just a style preference and if you all would rather have a property then I have no issues with that at all 🙂

Reply:
>  For me really the only important bit is to have a single API to get both and avoid doing two internal calls there 

If you optimize the `Target` property same way as `GCHandle.Target` as I have suggested in other comment, fetching Target won't be internal call anymore. And so the tuple returning method should not have a huge performance over just fetching `Target` and `Dependent` one at a time.

Reply:
Yup, was just going through the VM trying to understand exactly how that hack actually worked 😄
I wasn't aware that both `GCHandle` and `DependentHandle` still used the same `OBJECTHANDLE` struct in the VM (`GCHandleManager::CreateGlobalHandleOfType` and `GCHandleStore::CreateDependentHandle`). After going through the code and having a read at [this paragraph](https://github.com/dotnet/runtime/blob/01b7e73cd378145264a7cb7a09365b41ed42b240/docs/coding-guidelines/clr-code-guide.md#217-use-objecthandles-for-non-scoped-protection) I think I get the general gist of that seemingly magic trick, will go try to make that change you suggested. One never stops learning cool stuff by contributing to the runtime! 🙌

Reply:
@jkotas After checking it looks like that with Andy's recent jump threading changes (https://github.com/dotnet/runtime/pull/46257), the JIT can now correctly skip the second `null` check if you get `Target` and `Dependent` sequentially (given that `handle` just gets read to a register anyway), which after the optimization to `nGetPrimary` would've been the only remaining advantage for `GetTargetAndDependent()`. I think with that then the codegen for both things should be pretty much the same, so if you prefer to remove that API I guess it should be safe to do so now without having performance overhead for public consumers? Happy to add those remarks about `GC.KeepAlive` as well 🙂

Reply:
Yes, I would vote to remove the tuple returning APIs and just document the race conditions to be careful about.

Reply:
I still see it in the PR.  Did you intend to remove it?